Feature/1670 hide scroll view scroll bars

bmacombe opened the pull request:
### Description of Change ###

Added vertical and horizontal scroll bar visibility properties to ScrollView and updated UWP, WPF, iOS and Android renderers

I noticed on Android when the ScrollView.Orientation = ScrollOrientation.Both that a horizontal scroll bar is never shown, only the vertical.  I did not try to change this to avoid making a default behavior change.

Unit Tests are not included as I'm not sure what to test right now besides that basic properties were set.

### Bugs Fixed ###

- N/A

### API Changes ###

Added:

```
 public enum ScrollBarVisibility
 {
	Default,
	Always,
	Never,
} 
```
 - ScrollBarVisibility ScrollView.HorizontalScrollBarVisibility {get; set;};
 - ScrollBarVisibility ScrollView.VerticalScrollBarVisibility {get;set;};

### Behavioral Changes ###

Describe any non-bug related behavioral changes that may change how users app behaves when upgrading to this version of the codebase.

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

bmacombe committed:
Initial Commit

bmacombe committed:
Initial Commit

Added setting the initial state of H and V Scroll bar visibility on Android

Added WPF support

Fixed a copy/paste of Upw that should be Wpf

Added GTK and MacOS.

Plus a minor addition to iOS

Added Tizen support

I am unable to test as I can not get Tizen working in my envrionment at the moment

bmacombe committed:
Added setting the initial state of H and V Scroll bar visibility on Android

bmacombe commented on Xamarin.Forms.Platform.UAP/ScrollViewRenderer.cs:
```diff
@@ -4,6 +4,7 @@
 using Windows.Foundation;
 using Windows.UI.Xaml;
 using Windows.UI.Xaml.Controls;
+using UwpScrollBarVisibility = Windows.UI.Xaml.Controls.ScrollBarVisibility;
```
Needed to alias one of ScrollBarVisibility enums, decided to do that to the platform enum instead of forms

bmacombe commented on Xamarin.Forms.Platform.WPF/Renderers/EditorRenderer.cs:
```diff
@@ -2,6 +2,7 @@
 using System.Windows;
 using System.Windows.Controls;
 using System.Windows.Media;
+using WpfScrollBarVisibility = System.Windows.Controls.ScrollBarVisibility;
```
Needed to alias one of ScrollBarVisibility enums, decided to do that to the platform enum instead of forms

bmacombe committed:
Added WPF support

llevera commented on Xamarin.Forms.Platform.WPF/Renderers/ScrollViewRenderer.cs:
```diff
@@ -180,6 +188,28 @@ void UpdateScrollPosition()
 				Controller.SetScrolledPosition(Control.HorizontalOffset, Control.VerticalOffset);
 		}
 
+		WpfScrollBarVisibility ScrollBarVisibilityToUwp(ScrollBarVisibility visibility)
```
UWP in name a copy/paste?

llevera commented on Xamarin.Forms.Platform.WPF/Renderers/ScrollViewRenderer.cs:
```diff
@@ -156,14 +164,14 @@ void UpdateMargins()
 		void UpdateOrientation()
```
Won't this just throw away the scroll bar visibility set explicitly? Ie reorientation won't respect the enum

bmacombe replied:
I'll take a look at it again, I'll test changing the orientation and see how it affects things.

What about the case when you have the scrolling horizontal with the bar shown "Always" then switch to vertical.  Hide the bar like the current code does or honor the visibility  setting?

llevera commented on Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs:
```diff
@@ -424,5 +430,17 @@ void UpdateOrientation()
+
+		void UpdateHorizontalScrollBarVisibility()
+		{
+			if (_hScrollView != null)
```
Can use ?.

bmacombe replied:
Fine with me, doing the if seemed to fit in with the rest of the code style better me.

llevera replied:
Yeah, that seems fine if is matches current style

llevera commented on Xamarin.Forms.Platform.Android/Renderers/ScrollViewRenderer.cs:
```diff
@@ -424,5 +430,17 @@ void UpdateOrientation()
+
+		void UpdateVerticalScrollBarVisibility()
+		{
+			VerticalScrollBarEnabled = _view.VerticalScrollBarVisibility == ScrollBarVisibility.Always ||
```
What is `VerticalScrollBarEnabled` for and why not one for horizontal?

bmacombe replied:
Thats a good question, that was the suggested property to use over in the Forms issue on this.  I tried to use the Horizontal one, but it didn't do anything.

Seems to me they are nesting two Android scroll views to get vertical and horizontal scrolling.

llevera reviewed (changes requested):
Tizen, Mac, GTK? I can only assume we need to do these...?

bmacombe committed:
Fixed a copy/paste of Upw that should be Wpf

bmacombe commented:
@llevera Tizen, Mac, GTK?...I would guess we might, I don't have an environment to test those right now...I might be able to try the Mac one

llevera commented:
GTK and Mac are quite easy on Mac (I think). GTK just seems to run straight away and Mac only requires Xamarin.Mac

bmacombe commented:
I'll give it a try...never used any of the software on the Mac side...just use my MacBook Pro to run Win 10 in a VM and compile iOS software

llevera commented:
If you have hassles, I'm happy to do do those ones

Commit:
Added GTK and MacOS.

Plus a minor addition to iOS

bmacombe commented on Xamarin.Forms.Platform.GTK/Renderers/ScrollViewRenderer.cs:
```diff
@@ -34,8 +34,8 @@ protected override void OnElementChanged(ElementChangedEventArgs<ScrollView> e)
                         BorderWidth = 0,
-                        HscrollbarPolicy = PolicyType.Automatic,
-                        VscrollbarPolicy = PolicyType.Automatic
+                        HscrollbarPolicy = ScrollBarVisibilityToGtk(Element.HorizontalScrollBarVisibility),
```
As far as I can tell the GTK ScrolledWindow doesn't honer the Scrollbar policy.  Or some other action is required to get it to refresh after setting the value.  Someone who knows more about GTK might have to look at this.

bmacombe commented:
@llevera If you want to take a look at the GTK renderer please do.

I tried to get the Tizen SDK and emulator to work, and while I did the Control Gallery doesn't launch correctly in the Mobile Emulator.  I'll mess with it more, but if someone already has a Tizen environment working and wants to tackle that renderer, please feel free.

bmacombe committed:
Added Tizen support

I am unable to test as I can not get Tizen working in my envrionment at the moment

jsuarezruiz reviewed (dismissed):
I can review the behavior in GTK if you want, to see if I can help.

bmacombe commented:
@jsuarezruiz Please do, thanks

rookiejava commented:
As for Tizen, I can help to check the behavior. Just a quick scan of changes on Tizen/ScrollViewRenderer, looks good. And, it looks like you need a rebase onto latest upstream master.

bmacombe commented:
I think I rebased and squashed the commits...first time doing that, so I hope its right!

bmacombe commented:
I believe this PR is ready for review.  Verification of GTK and Tizen functionality is still needed.

GTK: Updated the renderers, but the GTK platform controls don't appear to honor the scroll bar visibility that is set.

Tizen: Tizen team thinks the code changes are good, just need to be verified by someone with a working Tizen environment.

Commit:
Add toolbox support to nuget package (#1871)

* Add toolbox support to nuget package

* Add picker/label to toolbox

hartez committed:
Don't assume casts to Activity from Context will be successful (#1867)

Commit:
Revert "Add toolbox support to nuget package (#1871)"

This reverts commit 0a2e52eb27d03d3567fd5b182013270d42a86d45.

jassmith commented on Xamarin.Forms.Core/ScrollView.cs:
```diff
@@ -26,6 +26,10 @@ public class ScrollView : Layout, IScrollViewController, IElementConfiguration<S
 
 		readonly Lazy<PlatformConfigurationRegistry<ScrollView>> _platformConfigurationRegistry;
 
+		public static readonly BindableProperty HorizontalScrollBarVisibilityProperty = BindableProperty.Create("HorizontalScrollBarVisibility", typeof(ScrollBarVisibility), typeof(ScrollView), ScrollBarVisibility.Default);
```
Consider using `nameof(HorizontalScrollBarVisibility)` instead of a string here to make refactoring cleaner if needed. Also helps avoiding typos or at least makes sure all typos are the same :P

bmacombe replied:
I normally do in my own code, I'm glad to change it, I was trying to keep the style similar to others I saw in the files for consistency. @jassmith   Would you like me to update them all to use nameof while I'm at it? I thought about it, but didn't want to change items that were not related.

jassmith replied:
nameof wasn't always there in the old days :) All new properties should use it. I have no problem if you want to submit a separate PR to fix all of them :P

jassmith commented on Xamarin.Forms.Platform.GTK/Renderers/ScrollViewRenderer.cs:
```diff
@@ -196,5 +202,27 @@ private void UpdateContentSize()
+        {
+            switch(visibility)
+            {
+                case ScrollBarVisibility.Default: return PolicyType.Automatic;
```
please put returns on their own lines.

jassmith commented on Xamarin.Forms.Platform.GTK/Renderers/ScrollViewRenderer.cs:
```diff
@@ -196,5 +202,27 @@ private void UpdateContentSize()
+
+        void UpdateHorizontalScrollBarVisibility()
+        {
+            if (Element.Orientation == ScrollOrientation.Horizontal || Element.Orientation == ScrollOrientation.Both)
```
`var orientation = Element.Orientation;` then use `orientation` wherever you need it. This reduces the number of calls to BindableObject.GetValue which is effectively a list lookup.

jassmith commented on Xamarin.Forms.Platform.MacOS/Renderers/ScrollViewRenderer.cs:
```diff
@@ -68,6 +68,8 @@ public void SetElement(VisualElement element)
 
 				UpdateContentSize();
 				UpdateBackgroundColor();
+                UpdateVerticalScrollBarVisibility();
```
whitespace

jassmith commented on Xamarin.Forms.Platform.MacOS/Renderers/ScrollViewRenderer.cs:
```diff
@@ -155,6 +157,10 @@ protected virtual void OnElementPropertyChanged(object sender, PropertyChangedEv
 				UpdateContentSize();
 			else if (e.PropertyName == VisualElement.BackgroundColorProperty.PropertyName)
 				UpdateBackgroundColor();
+            else if (e.PropertyName == ScrollView.VerticalScrollBarVisibilityProperty.PropertyName)
```
more indent issues

jassmith commented on Xamarin.Forms.Platform.MacOS/Renderers/ScrollViewRenderer.cs:
```diff
@@ -202,6 +208,22 @@ void UpdateContentSize()
 			ContentView.Frame = new RectangleF(ContentView.Frame.X, ContentView.Frame.Y, Frame.Width, Frame.Height);
 		}
 
+        void UpdateVerticalScrollBarVisibility()
```
and here, going to stop calling these out here

jassmith commented on Xamarin.Forms.Platform.Tizen/Renderers/ScrollViewRenderer.cs:
```diff
@@ -179,5 +185,27 @@ void OnScrollRequested(object sender, ScrollToRequestedEventArgs e)
+
+		void UpdateHorizontalScrollBarVisibility()
+		{
+			if (Element.Orientation == ScrollOrientation.Horizontal || Element.Orientation == ScrollOrientation.Both)
```
GetValue calls again here are being overused

jassmith commented on Xamarin.Forms.Platform.Tizen/Renderers/ScrollViewRenderer.cs:
```diff
@@ -179,5 +185,27 @@ void OnScrollRequested(object sender, ScrollToRequestedEventArgs e)
+		{
+			switch (visibility)
+			{
+				case ScrollBarVisibility.Default: return ScrollBarVisiblePolicy.Auto;
```
Formatting for case statements :)

jassmith commented on Xamarin.Forms.Platform.UAP/ScrollViewRenderer.cs:
```diff
@@ -188,12 +197,34 @@ void UpdateOrientation()
+
+		void UpdateHorizontalScrollBarVisibility()
+		{
+			if (Element.Orientation == ScrollOrientation.Horizontal || Element.Orientation == ScrollOrientation.Both)
```
GetValues here too

jassmith commented on Xamarin.Forms.Platform.WPF/Renderers/ScrollViewRenderer.cs:
```diff
@@ -180,6 +188,28 @@ void UpdateScrollPosition()
+
+		void UpdateHorizontalScrollBarVisibility()
+		{
+			if (Element.Orientation == ScrollOrientation.Horizontal || Element.Orientation == ScrollOrientation.Both)
```
And here

jassmith commented on Xamarin.Forms.Platform.iOS/Renderers/ScrollViewRenderer.cs:
```diff
@@ -181,6 +187,22 @@ void UpdateIsEnabled()
 
+		void UpdateVerticalScrollBarVisibility()
+		{
+			if (ScrollView.VerticalScrollBarVisibility == ScrollBarVisibility.Always || ScrollView.VerticalScrollBarVisibility == ScrollBarVisibility.Default)
```
`ShowVerticalScrolIndicator = ScrollView.VerticalScrollBarVisibility == ScrollBarVisibility.Always || ScrollView.VerticalScrollBarVisibility == ScrollBarVisibility.Default;`

jassmith commented on Xamarin.Forms.Platform.iOS/Renderers/ScrollViewRenderer.cs:
```diff
@@ -181,6 +187,22 @@ void UpdateIsEnabled()
+
+		void UpdateHorizontalScrollBarVisibility()
+		{
+			if (ScrollView.HorizontalScrollBarVisibility == ScrollBarVisibility.Always)
```
Same as above

jassmith reviewed (commented):
Lovely work, the GetValue calls should be fixed for performance reasons, but looks good to me.

bmacombe commented:
Thanks @jassmith I'll try to get that cleaned up this weekend and I've ask for some help on getting the PR moved over to the main branch.

bmacombe committed:
Changes from review

bmacombe committed:
Update ScrollViewRenderer.cs

try0 committed:
Fix to absolute URL

rmarinho committed:
[Build] Fix windows cert

rmarinho committed:
[Build]Fix master and build (#1920)

* [Build]Fix master and build

* [Pages] Fix certs

jfversluis committed:
Implemented MaxLength property on Entry and Editor (#1880)

* Implemented UAP and WPF

* Revert "Implemented UAP and WPF"

This reverts commit e17a66236756021e2b66503281ee3c1cf463e5da.

* Started implementation of #1663

Added MaxLength property on InputView and implemented iOS and Android
Entry and Android Editor

* Implemented iOS Editor

* Improved Android MaxLength

First check to see if there is a LengthFilter in there already and
remove that first

* Implemented GTK naming enhancement for iOS

* Implemented UAP and WPF

* Implemented Tizen

* Removed Linq and implemented forgotten method 😅

* Fixed whitespaces -> tabs

Except GTK, seems everything is spaces there, so kept it for consistency

* Added null guard for iOS

* Implemented Mac OS

* Added samples to gallery

* Polished Tizen implementation

* Tizen EditorRenderer updated and converted more whitespaces to tabs

* Fixed spaces to tabs and usage of nameof

* Added trimming of current text when MaxLength is less than current value length

* Reference right Entry for Tizen

* Trimming on MaxLength change for Tizen

* Trimming on MaxLength change for GTK

* Removed redundant GetValue calls and whitespace fiesta

* And the ones I missed

* Updated the docs

* Revert "Updated the docs"

This reverts commit 416e28706ab921d72e72ebc6488d4b3c4651a384.

* Updated docs just for InputView

rmarinho committed:
[UITest] Fix test for UITest package update (#1923)

alanag13 committed:
add small WPF performance improvement

alanag13 committed:
add docs

kzu committed:
Use relative URL to support recursive checkout in VSTS (#1926)

This is the norm in all other repos building in VSTS and
makes upstream build definitions and scripts much simpler.

rmarinho committed:
[iOS] ViewDidLayoutSubviews after removing page (#1532) fixes #1426

* [Controls] Repo for issue #1426

* [Controls] Simplify reproducion

* [iOS] Make sure to call layout when page is removed from stack and CurrentPage is updated

* [Controls] Fix issue number

* Update test and packages

* [Build] Fix windows cert

* [Build]Fix master and build (#1920)

* [Build]Fix master and build

* [Pages] Fix certs

* Update Xamarin.Forms.ControlGallery.WindowsUniversal.csproj

* Fix test again

* Update Bugzilla37841.cs

rmarinho committed:
[Android]?Remove UserVisibleHint (#1550) fixes #1438

rmarinho committed:
Merge branch '15-5'

alanag13 committed:
fix bad merge

rookiejava committed:
Support CascadeInputTransparent to Tizen (#1916)

henricm committed:
Added support for ListView full width separators on iOS (#1854) fixes #1665

* Fixes #1665 by adding `SeparatorStyle` property
to iOS specific ListViews. ListViewRenderer will check
this property when cells are retrieved in the ListViewDataSource
and set SeparatorInset and LayoutMargins to zero.

Commit:
[iOS] Add shadow effect (#1896)

* Add shadow effect for iOS

* Fix docs

* Fix formatting issues

davidortinau committed:
Merge pull request #1917 from try0/fix-CONTRIBUTING.md

[README] Fix to absolute URL

PureWeen committed:
set csharp_space_after_keywords_in_control_flow_statements to true to fit our design guide lines (#1964)

jcmanke committed:
[iOS Maps] Pin rendering customization (#1065)

* Changed MapRenderer to pure weak delegate pattern. GetViewForAnnotation
is virtual for pin rendering customization by subclasses.

* Added UI test for pin tap gesture

* Wait for callout before tapping it in UI test

kzu committed:
Simplify event raising invocation pattern (#1971)

This is effectively equivalent to the original code, yet nicer to read :)

This is also how it's being done elsewhere (i.e. fast renderers,
BindableObject, etc.), probably because that code is newer?

alanag13 committed:
light refactor (macOS)

alanag13 committed:
spaces -> tabs

alanag13 committed:
more spaces -> tabs

alanag13 committed:
account for Android styles when setting default behavior
